File: mock-server-java/src/main/java/com/example/aichat/service/ChatService.java
package com.example.aichat.service;

import com.example.aichat.config.AiChatProperties;
import com.example.aichat.dto.ChatRequest;
import com.example.aichat.dto.ChatStreamEvent;
import com.example.aichat.entity.Conversation;
import com.example.aichat.entity.Message;
import com.example.aichat.provider.ProviderRegistry;
import com.example.aichat.repository.MessageRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.data.domain.PageRequest;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.publisher.Sinks;
import reactor.core.scheduler.Schedulers;

import java.time.Instant;
import java.time.ZoneId;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicReference;

/**
 * 聊天核心编排 Service（对应 Node 版 services/chatService.js）
 * <p>职责：DB 读取历史 → 调 Provider → 流式转发 → DB 持久化</p>
 *
 * <p>WebFlux 注意事项：</p>
 * <ul>
 *   <li>JPA 是阻塞 IO，必须在 boundedElastic 线程池执行，不能阻塞 EventLoop</li>
 *   <li>Flux 取消（前端断开）会自动传播到上游 Provider 的 WebClient 调用</li>
 * </ul>
 */
@Slf4j
@Service
@RequiredArgsConstructor
public class ChatService {

    private static final DateTimeFormatter FORMATTER =
            DateTimeFormatter.ofPattern("yyyy/M/d HH:mm:ss").withZone(ZoneId.systemDefault());

    private final MessageRepository msgRepo;
    private final ProviderRegistry providerRegistry;
    private final AiChatProperties props;
    private final ChatPersistenceService persistence;   // 所有 DB 写操作代理到此 Bean，确保 @Transactional 生效

    /**
     * 处理一次聊天请求，返回 Flux 事件流
     * <p>Flux 中的事件类型：Meta → Chunk... → Finish → Done（或 Error）</p>
     */
    public Flux<ChatStreamEvent> handleChat(ChatRequest req, boolean mockError) {
        if (mockError) {
            return Flux.just(
                    new ChatStreamEvent.Error("MOCK_ERROR", "这是一条模拟的错误响应"),
                    new ChatStreamEvent.Done()
            );
        }

        // 使用 Sinks.Many 作为事件总线，在阻塞线程里 emit，在 Flux 里消费
        Sinks.Many<ChatStreamEvent> sink = Sinks.many().unicast().onBackpressureBuffer();

        // 所有 DB 操作 + Provider 调用放在 boundedElastic 线程池，不阻塞 EventLoop
        // 使用 Mono.fromRunnable：任务只需要被执行，不关心返回值
        Mono.fromRunnable(() -> doHandleChat(req, sink))
                .subscribeOn(Schedulers.boundedElastic())
                .subscribe(
                        v -> {},
                        err -> {
                            log.error("[chat] unhandled error: {}", err.getMessage(), err);
                            sink.tryEmitNext(new ChatStreamEvent.Error("INTERNAL_ERROR", err.getMessage()));
                            sink.tryEmitComplete();
                        }
                );

        return sink.asFlux();
    }

    /**
     * 实际编排逻辑（在 boundedElastic 线程执行）
     */
    private void doHandleChat(ChatRequest req, Sinks.Many<ChatStreamEvent> sink) {
        String model = req.model() == null ? "gpt-4o" : req.model();

        // 1. 解析 / 创建会话
        Conversation conv = persistence.resolveOrCreate(
                req.conversationId(), truncateTitle(req.content()), model, req.systemPrompt());

        // 2. 写入用户消息
        Message userMsg = persistence.saveMessage(conv.getId(), "user", req.content(), null, req.attachments());

        // 3. 创建占位 assistant 消息
        Message assistantMsg = persistence.saveMessage(conv.getId(), "assistant", "", model, null);

        // 4. 路由 Provider
        ProviderRegistry.Resolved resolved = providerRegistry.resolve(model);
        log.info("[chat] conv={} model={} provider={} fallback={}",
                conv.getId(), model, resolved.provider().name(), resolved.fallback());

        // 5. 推送 meta 首帧
        sink.tryEmitNext(new ChatStreamEvent.Meta(
                conv.getId(), conv.getTitle(),
                userMsg.getId(), assistantMsg.getId(),
                resolved.provider().name(), resolved.requestedProvider(),
                resolved.fallback(), resolved.realModel()
        ));

        // 6. 组装上下文
        List<Map<String, String>> messages = buildContext(conv, req);

        // 7. 调用 Provider 流式接口
        String chunkId = "chatcmpl-" + Long.toHexString(System.currentTimeMillis());
        long created = System.currentTimeMillis() / 1000;
        AtomicReference<String> fullContent = new AtomicReference<>("");

        Map<String, Object> params = new HashMap<>();
        if (req.temperature() != null) params.put("temperature", req.temperature());
        if (req.maxTokens() != null) params.put("maxTokens", req.maxTokens());
        if (req.topP() != null) params.put("topP", req.topP());

        try {
            resolved.provider()
                    .streamChat(messages, resolved.realModel(), params)
                    .doOnNext(text -> {
                        fullContent.updateAndGet(prev -> prev + text);
                        sink.tryEmitNext(new ChatStreamEvent.Chunk(text, chunkId, created, model));
                    })
                    .doOnComplete(() -> {
                        // 写入 finish chunk
                        sink.tryEmitNext(new ChatStreamEvent.Finish("stop", chunkId, created, model));
                        // 持久化 assistant 消息
                        persistence.finalizeAssistantMessage(assistantMsg.getId(), fullContent.get(), "stop", null);
                        persistence.touchConversation(conv.getId());
                        sink.tryEmitNext(new ChatStreamEvent.Done());
                        sink.tryEmitComplete();
                    })
                    .doOnError(err -> {
                        log.warn("[chat] provider error: {}", err.getMessage());
                        persistence.finalizeAssistantMessage(assistantMsg.getId(), fullContent.get(), "error", err.getMessage());
                        persistence.touchConversation(conv.getId());
                        sink.tryEmitNext(new ChatStreamEvent.Error("PROVIDER_ERROR",
                                "模型调用失败：" + err.getMessage()));
                        sink.tryEmitNext(new ChatStreamEvent.Done());
                        sink.tryEmitComplete();
                    })
                    .blockLast();   // 在 boundedElastic 线程阻塞等待流完成
        } catch (Exception e) {
            if (!Thread.currentThread().isInterrupted()) {
                log.warn("[chat] stream exception: {}", e.getMessage());
                persistence.finalizeAssistantMessage(assistantMsg.getId(), fullContent.get(), "error", e.getMessage());
                persistence.touchConversation(conv.getId());
                sink.tryEmitNext(new ChatStreamEvent.Error("STREAM_ERROR", e.getMessage()));
                sink.tryEmitNext(new ChatStreamEvent.Done());
                sink.tryEmitComplete();
            } else {
                // 前端中断（AbortError），保存已生成部分
                persistence.finalizeAssistantMessage(assistantMsg.getId(), fullContent.get(), "abort", null);
                persistence.touchConversation(conv.getId());
                sink.tryEmitComplete();
            }
        }
    }

    // ---- 私有辅助方法 ----

    private List<Map<String, String>> buildContext(Conversation conv, ChatRequest req) {
        List<Map<String, String>> messages = new ArrayList<>();
        String system = req.systemPrompt() != null ? req.systemPrompt() : conv.getSystemPrompt();
        if (system != null && !system.isBlank()) {
            messages.add(Map.of("role", "system", "content", system));
        }
        // 取最近 N 条（含本轮 user 消息）
        List<Message> recent = msgRepo.findRecent(conv.getId(),
                PageRequest.of(0, props.contextMessageLimit()));
        // findRecent 是 DESC，需要反转
        for (int i = recent.size() - 1; i >= 0; i--) {
            Message m = recent.get(i);
            if (!"system".equals(m.getRole()) && !m.getContent().isBlank()) {
                messages.add(Map.of("role", m.getRole(), "content", m.getContent()));
            }
        }
        return messages;
    }

    private static String truncateTitle(String text) {
        if (text == null || text.isBlank()) return "新会话";
        String t = text.replaceAll("\\s+", " ").trim();
        return t.length() > 20 ? t.substring(0, 20) + "..." : t;
    }

    public String formatTime(long ts) {
        return FORMATTER.format(Instant.ofEpochMilli(ts));
    }
}

File: mock-server-node/services/chatService.js
/**
 * 对话 Service
 * @description 编排：DB 读取历史 → 调 Provider → 流式转发 → DB 持久化
 */
import {
	createConversation,
	getConversation,
	touchConversation,
	updateConversation,
} from '../db/dao/conversations.js';
import {
	createMessage,
	finalizeMessage,
	getRecentMessages,
} from '../db/dao/messages.js';
import { resolveProvider } from '../providers/index.js';

const CONTEXT_LIMIT = Number(process.env.CONTEXT_MESSAGE_LIMIT) || 20;

/**
 * 处理一次对话请求
 *
 * @param {object} payload - 前端请求体
 * @param {string} [payload.conversationId] - 会话 id，为空则自动创建
 * @param {string} payload.content - 用户输入
 * @param {string} payload.model
 * @param {Array} [payload.attachments]
 * @param {string} [payload.systemPrompt]
 * @param {number} [payload.temperature]
 * @param {number} [payload.maxTokens]
 *
 * @param {object} handlers
 * @param {(meta:object)=>void} handlers.onMeta  - 首帧元数据（含 conversationId/messageId/provider）
 * @param {(text:string)=>void} handlers.onChunk
 * @param {(result:object)=>void} handlers.onDone
 * @param {(err:Error)=>void} handlers.onError
 * @param {AbortSignal} handlers.signal
 */
export const handleChat = async (payload, handlers) => {
	const {
		conversationId: convIdInput,
		content,
		model = 'gpt-4o',
		attachments = [],
		systemPrompt,
		temperature,
		maxTokens,
	} = payload;
	const { onMeta, onChunk, onDone, onError, signal } = handlers;

	// 1. 解析 / 创建会话
	let conversation = convIdInput ? getConversation(convIdInput) : null;
	if (!conversation) {
		conversation = createConversation({
			title: truncateTitle(content),
			model,
			systemPrompt,
		});
	}

	// 2. 写入用户消息
	const userMsg = createMessage({
		conversationId: conversation.id,
		role: 'user',
		content,
		attachments,
	});

	// 3. 创建占位 assistant 消息（流式过程中追加内容）
	const assistantMsg = createMessage({
		conversationId: conversation.id,
		role: 'assistant',
		content: '',
		model,
	});

	// 4. 解析 Provider
	const { provider, realModel, fallback, requestedProvider } = resolveProvider(model);

	// 5. 推送元数据帧（前端可据此更新 conversationId / 刷新侧边栏）
	onMeta?.({
		conversationId: conversation.id,
		conversationTitle: conversation.title,
		userMessageId: userMsg.id,
		assistantMessageId: assistantMsg.id,
		provider: provider.name,
		requestedProvider,
		fallback,
		realModel,
	});

	// 6. 组装上下文
	const history = getRecentMessages(conversation.id, CONTEXT_LIMIT);
	const messages = [];
	const effectiveSystem = systemPrompt || conversation.systemPrompt;
	if (effectiveSystem) {
		messages.push({ role: 'system', content: effectiveSystem });
	}
	// history 已包含最新的 user 消息（含本轮）
	for (const m of history) {
		if (m.role !== 'system') {
			messages.push({ role: m.role, content: m.content });
		}
	}

	// 7. 流式调用 Provider，同时累积全文以便最终入库
	let fullContent = '';
	let providerError = null;

	await provider.streamChat({
		messages,
		model: realModel,
		temperature,
		maxTokens,
		signal,
		onChunk: (text) => {
			fullContent += text;
			onChunk?.(text);
		},
		onDone: (meta) => {
			const aborted = meta?.aborted || signal?.aborted;
			const finishReason = aborted ? 'abort' : 'stop';

			finalizeMessage(assistantMsg.id, {
				content: fullContent,
				tokensOut: meta?.tokensOut || 0,
				finishReason,
				error: null,
			});
			touchConversation(conversation.id);

			// 首轮对话自动用回复内容优化标题
			if (conversation.title === '新会话' && fullContent) {
				updateConversation(conversation.id, { title: truncateTitle(content) });
			}

			onDone?.({
				conversationId: conversation.id,
				assistantMessageId: assistantMsg.id,
				finishReason,
				contentLength: fullContent.length,
			});
		},
		onError: (err) => {
			providerError = err;
			finalizeMessage(assistantMsg.id, {
				content: fullContent,
				finishReason: 'error',
				error: String(err?.message || err),
			});
			touchConversation(conversation.id);
			onError?.(err);
		},
	});

	return { error: providerError };
};

/**
 * 将用户首条消息截取为标题（最多 20 字）
 */
const truncateTitle = (text) => {
	const t = String(text || '').replace(/\s+/g, ' ').trim();
	if (!t) return '新会话';
	return t.length > 20 ? `${t.slice(0, 20)}...` : t;
};
